docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+`// +default=custom` marker will force a stub to be generated. Stubs will be
+generated with a `panic("write me!")` marker to ensure that they are populated.
+
+### Package: `apis/group/v1/install/validation`
+
+This package contains validation functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+The allowed imports are constrained; in particular, nothing that would let you
+do RPCs to look up other cluster state is allowed. Validation functions validate
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere. (TODO: copy-pasta this 2x above)
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validate_min=<numeric value>,validate_max=<numeric value>`
+* `// +validate_regexp=<regexp>`
+* `// +validate_exact=<exact value>`
+
+If no validation directive is in the comments, a stub with a
+`panic("write me!")` marker will be generated. To prevent this, a
+`// +no_validation_required` directive may be added to the comment.
+
+### Package: `apis/group/v1beta1`
+
+This package layout is the same as `apis/group/v1`, but it is an older version
+of the API group, which may lack features.
+
+### Package: `apis/group/v2alpha1`
+
+This package layout is the same as `apis/group/v1`, but it is an newer version
+of the API group, which may have extra features.
+
+## Type support package layout
+
+The above is how to arrange the types for a particular API. But we also must
+consider, what Kubernetes packages do you need to import to use the API types?
+
+```
+pkg
+└── typetools
+    ├── convert
+    ├── deep
+    ├── default
+    └── validate
+```
+
+These packages will live in the kubernetes repo (or another repo that the
+kubernetes team will own and maintain). The interface for all the packages is
+similar: there'll be a `Register(...)` function for the install packages to use,
+and some function to perform the action. So, e.g., `convert.Convert(&in, &out)`,
+`bar := deep.Copy(&foo)`, `default.Apply(&object)`,
+`errs := validate.Check(&object)`.
+
+More detailed descriptions belong in another doc when we hash out the
+particulars of the code generation; the important thing here is just to know
+that we'll supply packages with:
+* Minimal import tree
+* Some sort of registration method
+* Some sort of execution method
+
+## Reusable apiserver infrastructure packages
+
+WIP
+
+In the future, it'd be nice if we could generate REST install code based on
+comment annotations in the go types.
+
+## Desired toolchain
+
+Along with packages for runtime use, we will supply tools to do the various
+generations.
+
+Ideally, the tool will take the root `apis/` directory, discover the `group`s
```

Comment:
If we argued today that each of the resource types we have today is really an API group, that would just mean we'd have to give up our monolithic root API group.  That's not a bad thing (logically, we expected builds to be distinct from images), although until the infrastructure for groups gets better I wouldn't want to change it aggressively.  Since this is generation, if you didn't support that scenario we could simply live without it until our next iteration of APIs.

Of course, if the *output* is a single directory that's fine for us today.  It's really only the assumption that our types live in a particular package (vs being registered on the scheme) that causes us an issue.